kvserver: fail helpfully in and deflake TestOptimisticEvalNoContention

tbg opened the pull request:
We've seen[^1] that in rare cases, the limited scan in that test does
not successfully evaluate optimistically, and ends up blocking on an
intent indefinitely.

This commit adds a timeout to the scan as well as wraps it in a verbose
trace so that we can understand this problem better the next time we
see it.

I looked at the optimistic read path and see no obvious way this can
be turned off (short of randomizing the cluster setting[^2], which we
don't seem to be doing). So perhaps the read hits a retry or the like,
but either way it will be less poking around in the dark once we get
a repro after this commit.

[^1]: https://github.com/cockroachdb/cockroach/issues/123986#issuecomment-2107606634
[^2]: https://github.com/cockroachdb/cockroach/issues/123986#issuecomment-2107606634

Closes #130973.

Epic: none
Release note: None

tbg committed:
kvserver: fail helpfully in TestOptimisticEvalNoContention

We've seen[^1] that in rare cases, the limited scan in that test does
not successfully evaluate optimistically, and ends up blocking on an
intent indefinitely.

This commit adds a timeout to the scan as well as wraps it in a verbose
trace so that we can understand this problem better the next time we
see it.

I looked at the optimistic read path and see no obvious way this can
be turned off (short of randomizing the cluster setting[^2], which we
don't seem to be doing). So perhaps the read hits a retry or the like,
but either way it will be less poking around in the dark once we get
a repro after this commit.

[^1]: https://github.com/cockroachdb/cockroach/issues/123986#issuecomment-2107606634
[^2]: https://github.com/cockroachdb/cockroach/issues/123986#issuecomment-2107606634

Closes #130973.

Epic: none
Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/131071)
<!-- Reviewable:end -->

stevendanna commented on pkg/kv/kvserver/client_replica_test.go:
```diff
@@ -5259,7 +5263,8 @@ func TestOptimisticEvalNoContention(t *testing.T) {
 	defer log.Scope(t).Close(t)
 
-	ctx := context.Background()
+	ctx, cancel := context.WithTimeout(context.Background(), 15*time.Second)
+	defer cancel()
```
If you wanted, you could use `testutils.SucceedsSoonDuration()` for the timeout. It is 45 seconds by default though, but has the advantage of being scaled under race/deadlock.  Although, I imagine even under race/deadlock we expect this to take much less than 15 seconds.

nvb reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 1 of 1 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/131071#-:-O7UhUmS5HUWijxp86Et:bnfp4nl)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @tbg)



<!-- Sent from Reviewable.io -->

tbg commented:
I can't reproduce this locally (still figuring out how to effectively stress tests with bazel) but CI gets it every time 

```
             0.838ms      0.035ms                event:n1 scan stats: stepped 2 times (5 internal); seeked 2 times (2 internal); block-bytes: (total 0 B, cached 0 B, duration 0s); points: (count 6, key-bytes 126 B, value-bytes 180 B, tombstoned: 0) ranges: (count 0), (contained-points 0, skipped-points 0) evaluated requests: 0 gets, 1 scans, 0 reverse scans
             0.866ms      0.029ms                event:kv/kvserver/concurrency/concurrency_manager.go:784 [n1,s1,r5/1:/{Systemtse-Table/0}] latch conflict after optimistic eval
             0.937ms      0.071ms                event:kv/kvserver/replica_read.go:175 [n1,s1,r5/1:/{Systemtse-Table/0}] conflicts detected on latches read global: {a-b} at 1728049701.452916250,0
```

tbg commented:
I added a commit that (I think) commits the initial writes in the test setup via 1PC. The repro above still works, unfortunately, failing like this:

```
             0.871ms      0.390ms                event:kv/kvserver/replica_evaluate.go:511 [n1,s1,r5/1:/{Systemtse-Table/0}] evaluated Scan command header:<key:"a" end_key:"c" > , txn="unnamed" meta={id=7803693e key=/Min iso=Serializable pri=0.00905590 epo=0 ts=1730220290
.445205712,0 min=1730220290.445205712,0 seq=0} lock=false stat=PENDING rts=1730220290.445205712,0 wto=false gul=1730220290.945205712,0 : resp=header:<resume_span:<key:"b" end_key:"c" > resume_reason:RESUME_KEY_LIMIT num_keys:1 num_bytes:25 > rows:<key:"a" value:<raw_bytes
:"\247\314z\010\003a" timestamp:<wall_time:1730220290443925970 > > > col_batches:<> , err=<nil>
             0.915ms      0.045ms                event:n1 scan stats: stepped 1 times (1 internal); seeked 2 times (2 internal); block-bytes: (total 0 B, cached 0 B, duration 0s); points: (count 2, key-bytes 22 B, value-bytes 12 B, tombstoned: 0) ranges: (count 0), (conta
ined-points 0, skipped-points 0) evaluated requests: 0 gets, 1 scans, 0 reverse scans
             1.023ms      0.108ms                event:kv/kvserver/spanlatch/manager.go:356 [n1,s1,r5/1:/{Systemtse-Table/0}] {a-b}@1730220290.945205712,0 overlaps at a@1730220290.443925970,0 for request Put ["a"], Put ["b"], EndTxn(commit) ["a"], [txn: a92780d8], [can-fo
rward-ts]
             1.052ms      0.029ms                event:kv/kvserver/concurrency/concurrency_manager.go:778 [n1,s1,r5/1:/{Systemtse-Table/0}] latch conflict after optimistic eval
             1.182ms      0.130ms                event:kv/kvserver/replica_read.go:170 [n1,s1,r5/1:/{Systemtse-Table/0}] conflicts detected on latches read global: {a-b} at 1730220290.945205712,0
        read local: /Local/RangeID/5/r/AbortSpan/"7803693e-8e13-4448-bafa-1a23f570904b" at 0,0
         and locks None: {a-b}
        ; transforming into optimistic eval error
             1.265ms      0.082ms                event:kv/kvserver/replica_send.go:507 [n1,s1,r5/1:/{Systemtse-Table/0}] concurrency retry error: optimistic eval encountered conflict
```

In other word, we now block for the latch for `Put(a)Put(b)EndTxn(a)`, despite having previously read from both `a` and `b`.

tbg commented:
More confusion. Example from the trace - we block on the 1PC commit, or rather its latch shows up. Note the `evaluated ...` line, which comes from `--vmodule=replica_evaluate=2`.

```
             0.756ms      0.193ms                event:kv/kvserver/replica_evaluate.go:511 [n1,s1,r5/1:/{Systemtse-Table/0}] evaluated Scan command header:<key:"a" end_key:"c" > , txn="unnamed" meta={id=cf3e15f5 key=/Min iso=Serializable pri=0.01498624 epo=0 ts=1730305914.697938221,0 min=1730305914.697938221,0 seq=0} lock=false stat=PENDING rts=1730305914.697938221,0 wto=false gul=1730305915.197938221,0 : resp=header:<resume_span:<key:"b" end_key:"c" > resume_reason:RESUME_KEY_LIMIT num_keys:1 num_bytes:25 > rows:<key:"a" value:<raw_bytes:"\247\314z\010\003a" timestamp:<wall_time:1730305914694830531 > > > col_batches:<> , err=<nil>
             0.801ms      0.045ms                event:n1 scan stats: stepped 1 times (1 internal); seeked 2 times (2 internal); block-bytes: (total 0 B, cached 0 B, duration 0s); points: (count 2, key-bytes 22 B, value-bytes 12 B, tombstoned: 0) ranges: (count 0), (contained-points 0, skipped-points 0) evaluated requests: 0 gets, 1 scans, 0 reverse scans
             0.879ms      0.078ms                event:kv/kvserver/spanlatch/manager.go:356 [n1,s1,r5/1:/{Systemtse-Table/0}] {a-b}@1730305915.197938221,0 overlaps at a@1730305914.694830531,0 for request Put ["a"], Put ["b"], EndTxn(commit) ["a"], [txn: 67644d75], [can-forward-ts]
             0.896ms      0.017ms                event:kv/kvserver/concurrency/concurrency_manager.go:778 [n1,s1,r5/1:/{Systemtse-Table/0}] latch conflict after optimistic eval
```

We can look at all of the `evaluated ...` messages to get a sequence of global events, see below. We see two puts (1pc, so no EndTxn), then the gets for a and b, then our failing optimistic scan. This doesn't make sense, if the gets go through, then the 1PC txn must have released latches already, and then how can we get stuck on latches if there isn't anything else coming in in the meantime to grab them?

```
[...]
I241030 16:31:54.695086 66 kv/kvserver/replica_evaluate.go:511 ⋮ [T1,Vsystem,n1,s1,r5/1:‹/{Systemtse-Table/0}›] 5418  evaluated Put command ‹header:<key:"a" sequence:1 > value:<raw_bytes:"\247\314z\010\003a" timestamp:<> > ›, txn=<nil> : resp=‹header:<> ›, err=<nil>
I241030 16:31:54.695154 66 kv/kvserver/replica_evaluate.go:511 ⋮ [T1,Vsystem,n1,s1,r5/1:‹/{Systemtse-Table/0}›] 5419  evaluated Put command ‹header:<key:"b" sequence:2 > value:<raw_bytes:"<\203\225\353\003b" timestamp:<> > ›, txn=<nil> : resp=‹header:<> ›, err=<nil>
I241030 16:31:54.697260 66 kv/kvserver/replica_evaluate.go:511 ⋮ [T1,Vsystem,n1,s1,r5/1:‹/{Systemtse-Table/0}›] 5420  evaluated Get command ‹header:<key:"a" > ›, txn=<nil> : resp=‹header:<num_keys:1 num_bytes:6 > value:<raw_bytes:"\247\314z\010\003a" timestamp:<wall_time:1730305914694830531 > > ›, err=<nil>
I241030 16:31:54.697488 66 kv/kvserver/replica_evaluate.go:511 ⋮ [T1,Vsystem,n1,s1,r5/1:‹/{Systemtse-Table/0}›] 5421  evaluated Get command ‹header:<key:"b" > ›, txn=<nil> : resp=‹header:<num_keys:1 num_bytes:6 > value:<raw_bytes:"<\203\225\353\003b" timestamp:<wall_time:1730305914694830531 > > ›, err=<nil>
I241030 16:31:54.697784 66 kv/kvserver/replica_evaluate.go:511 ⋮ [T1,Vsystem,n1,s1,r5/1:‹/{Systemtse-Table/0}›] 5422  evaluated Scan command ‹header:<key:"b" end_key:"c" > key_locking_strength:Exclusive ›, txn="locking txn" meta={id=0d1c1936 key=‹"b"› iso=Serializable pri=0.02418677 epo=0 ts=1730305914.697549950,0 min=1730305914.697549950,0 seq=0} lock=true stat=PENDING rts=1730305914.697549950,0 wto=false gul=1730305915.197549950,0 : resp=‹header:<num_keys:1 num_bytes:25 > rows:<key:"b" value:<raw_bytes:"<\203\225\353\003b" timestamp:<wall_time:1730305914694830531 > > > col_batches:<> ›, err=<nil>
I241030 16:31:54.698662 2187 kv/kvserver/replica_evaluate.go:511 ⋮ [T1,Vsystem,n1,s1,r5/1:‹/{Systemtse-Table/0}›] 5423  evaluated Scan command ‹header:<key:"a" end_key:"c" > ›, txn="unnamed" meta={id=cf3e15f5 key=/Min iso=Serializable pri=0.01498624 epo=0 ts=1730305914.697938221,0 min=1730305914.697938221,0 seq=0} lock=false stat=PENDING rts=1730305914.697938221,0 wto=false gul=1730305915.197938221,0 : resp=‹header:<resume_span:<key:"b" end_key:"c" > resume_reason:RESUME_KEY_LIMIT num_keys:1 num_bytes:25 > rows:<key:"a" value:<raw_bytes:"\247\314z\010\003a" timestamp:<wall_time:1730305914694830531 > > > col_batches:<> ›, err=<nil>
[100ms pass]
I241030 16:31:54.803387 2187 kv/kvserver/replica_evaluate.go:511 ⋮ [T1,Vsystem,n1,s1,r5/1:‹/{Systemtse-Table/0}›] 5424  evaluated PushTxn command ‹header:<key:"b" > pusher_txn:<meta:<id:cf3e15f5-8102-4daf-a67d-178079b31a10 write_timestamp:<wall_time:1730305914697938221 > min_timestamp:<wall_time:1730305914697938221 > priority:321827 coordinator_node_id:1 > name:"unnamed" last_heartbeat:<wall_time:...›, txn=<nil> : resp=‹header:<> pushee_txn:<meta:<id:0d1c1936-304b-41a5-ae81-007dee15972b key:"b" write_timestamp:<wall_time:1730305914697549950 > min_timestamp:<wall_time:1730305914697549950 > priority:519407 coordinator_node_id:1 > last_heartbeat:<wall_time:173030591469754...›, err=failed to push meta={id=0d1c1936 key=‹"b"› iso=Serializable pri=0.02418677 epo=0 ts=1730305914.697549950,0 min=1730305914.697549950,0 seq=0} lock=true stat=PENDING rts=0,0 wto=false gul=0,0
I241030 16:31:54.804212 2187 kv/kvserver/replica_evaluate.go:511 ⋮ [T1,Vsystem,n1,s1,r5/1:‹/{Systemtse-Table/0}›] 5425  evaluated QueryTxn command ‹header:<key:"b" > txn:<id:0d1c1936-304b-41a5-ae81-007dee15972b key:"b" write_timestamp:<wall_time:1730305914697549950 > min_timestamp:<wall_time:1730305914697549950 > priority:519407 coordinator_node_id:1 > ›, txn=<nil> : resp=‹header:<> queried_txn:<meta:<id:0d1c1936-304b-41a5-ae81-007dee15972b key:"b" write_timestamp:<wall_time:1730305914697549950 > min_timestamp:<wall_time:1730305914697549950 > priority:519407 coordinator_node_id:1 > last_heartbeat:<wall_time:17303059146975...›, err=<nil>
[...]
```

tbg commented:
It looks like this doesn't repro with `DisableCanAckBeforeReplication`. So the question becomes - how can `db.Get(a)` and `db.Get(b)` both succeed before the latches for the write have been released?
I'll try to repro for a little longer before I seriously consider it valid that this is happening.

tbg commented:
It does look as though somehow the latch for the write is released only way too late (even after the scan succeeded on the pessimistic path), and I don't understand how the previous `Get` requests can have succeeded. None of this makes sense to me.

```
08:09:45.555591 68 kv/kvserver/replica_evaluate.go:512 ⋮   evaluated Put command ‹header:<key:"a" sequence:1 > value:<raw_bytes:"\247\314z\010\003a" timestamp:<> > ›, txn=<nil> : resp=‹header:<> ›, err=<nil>
08:09:45.555652 68 kv/kvserver/replica_evaluate.go:512 ⋮   evaluated Put command ‹header:<key:"b" sequence:2 > value:<raw_bytes:"<\203\225\353\003b" timestamp:<> > ›, txn=<nil> : resp=‹header:<> ›, err=<nil>
--- 1pc txn returned ---
08:09:45.556455 68 kv/kvserver/replica_evaluate.go:512 ⋮   evaluated Get command ‹header:<key:"a" > ›, txn=<nil> : resp=‹header:<num_keys:1 num_bytes:6 > value:<raw_bytes:"\247\314z\010\003a" timestamp:<wall_time:1730362185555306197 > > ›, err=<nil>
08:09:45.556657 68 kv/kvserver/replica_evaluate.go:512 ⋮   evaluated Get command ‹header:<key:"b" > ›, txn=<nil> : resp=‹header:<num_keys:1 num_bytes:6 > value:<raw_bytes:"<\203\225\353\003b" timestamp:<wall_time:1730362185555306197 > > ›, err=<nil>
^-- its writes were observed
08:09:45.556941 68 kv/kvserver/replica_evaluate.go:512 ⋮   evaluated Scan command ‹header:<key:"b" end_key:"c" > key_locking_strength:Exclusive ›, txn="locking txn" meta={id=fc44025d key=‹"b"› iso=Serializable pri=0.06035259 epo=0 ts=1730362185.556710654,0 min=1730362185.556710654,0 seq=0} lock=true stat=PENDING rts=1730362185.556710654,0 wto=false gul=1730362186.056710654,0 : resp=‹header:<num_keys:1 num_bytes:25 > rows:<key:"b" value:<raw_bytes:"<\203\225\353\003b" timestamp:<wall_time:1730362185555306197 > > > col_batches:<> ›, err=<nil>
^-- optimistic eval attempt, fails on latches
08:09:45.557670 2207 kv/kvserver/replica_evaluate.go:512 ⋮ evaluated Scan command ‹header:<key:"a" end_key:"c" > ›, txn="unnamed" meta={id=f6eb72ed key=/Min iso=Serializable pri=0.00559064 epo=0 ts=1730362185.557067739,0 min=1730362185.557067739,0 seq=0} lock=false stat=PENDING rts=1730362185.557067739,0 wto=false gul=1730362186.057067739,0 : resp=‹header:<resume_span:<key:"b" end_key:"c" > resume_reason:RESUME_KEY_LIMIT num_keys:1 num_bytes:25 > rows:<key:"a" value:<raw_bytes:"\247\314z\010\003a" timestamp:<wall_time:1730362185555306197 > > > col_batches:<> ›, err=<nil>
^-- pessimistic eval attempt
08:09:45.561819 485 kv/kvserver/replica_proposal.go:258 ⋮  TBG released latch for ‹Put ["a"], Put ["b"], EndTxn(commit) ["a"], [txn: 1b6f1315], [can-forward-ts]›
^-- latch is released
08:09:45.659324 2207 kv/kvserver/replica_evaluate.go:512 ⋮ evaluated PushTxn command ‹header:<key:"b" > pusher_txn:<meta:<id:f6eb72ed-a320-4610-83b1-20040e8c7110 write_timestamp:<wall_time:1730362185557067739 > min_timestamp:<wall_time:1730362185557067739 > priority:120058 coordinator_node_id:1 > name:"unnamed" last_heartbeat:<wall_time:...›, txn=<nil> : resp=‹header:<> pushee_txn:<meta:<id:fc44025d-1656-4384-990d-6bdbb0f83a72 key:"b" write_timestamp:<wall_time:1730362185556710654 > min_timestamp:<wall_time:1730362185556710654 > priority:1296062 coordinator_node_id:1 > last_heartbeat:<wall_time:17303621855567...›, err=failed to push meta={id=fc44025d key=‹"b"› iso=Serializable pri=0.06035259 epo=0 ts=1730362185.556710654,0 min=1730362185.556710654,0 seq=0} lock=true stat=PENDING rts=0,0 wto=false gul=0,0
```

tbg commented:
Okay I think I've got it. Reminder, we're seeing the following:

- `Put(a)Put(b)Commit` is run as a 1PC txn (in my current version of the failing test; it's not 1PC in the original test but it shouldn't mattern).
- when this returns, we run `db.Get(a)`, `db.Get(b)`.
- when this returns, we run `Scan(a,c,limit=1)` and this first takes the optimistic path, but then checks latches and **finds the latch for `Put(a)Put(b)Commit` still in place**.
- this fails the test (simplified but true).

The highlighted bit is what we need to understand. Here's how this can happen:

- `Put(a)Put(b)Commit` runs (1PC). Because we're acking writes before they apply (pipelined writes), this returns while the write is still inflight. I don't think this is necessary, but it probably helps actually hitting the right interleaving (when I turn it off, I can't repro).
- our `db.Get(a)` can race with the 1PC txn finishining application. Concretely, the 1PC can be in the middle of its call to `ec.repl.concMgr.FinishReq`[^1].
- concretely, in the middle of `(*spanlatch.Manager).Release`[^2].
- more concretely, it may already have called `lg.done.signal()`, but not yet `m.remove(ctx, lg)`.
- because `lg.done` has been signaled, `db.Get(a)` and `db.Get(b)` can go through and read the correct write.
- `Scan(a,c,limit=1)` arrives and optimistically evaluates. Then it checks latches - they are still there.
- at some point - in fact momentarily - `m.remove(ctx, lg)` will be called, but it is too late.
- the scan flunked out of the optimistic path.

If I replace the order of `m.remove` and `lg.done.signal`, this problem cannot occur. The test is stable. If I don't reorder but put in a well-timed sleep, the test fails deterministically:

```go
func (m *Manager) Release(ctx context.Context, lg *Guard) {
	lg.done.signal()
	if lg.snap != nil {
		lg.snap.close()
	}
	if ctx.Value(5) != nil {
		time.Sleep(100 * time.Millisecond)
	}
	m.remove(ctx, lg)
```

@nvanbenschoten are we okay reordering these or is this a performance optimization that we want to keep? Something else we could do to address this is to never bail on latches that have already been signaled. Concretely, in `CheckOptimisticNoConflicts`[^1] we'd modify the `overlaps` method to ignore latches that have `it.Cur().g.done.signaled()`.

[^2]: https://github.com/cockroachdb/cockroach/blob/51de177f88011d8e29e103973c1a8d17aa1dc947/pkg/kv/kvserver/spanlatch/manager.go#L641-L661
[^1]: https://github.com/cockroachdb/cockroach/blob/51de177f88011d8e29e103973c1a8d17aa1dc947/pkg/kv/kvserver/replica_send.go#L1360

tbg committed:
kvserver: deflake TestOptimisticEvalNoContention

The setup of the initial writes could leave latches behind that had been
signaled but were not removed from the btrees yet. Whenever the
optimistic read would have these latches in its latch snapshot, it would
fail and fail the test.

We change the harness used by this and related tests to disable
ack-before-application, which implies that the latches are gone by the
time the transaction returns.

For the detailed analysis, see:
https://github.com/cockroachdb/cockroach/pull/131071#issuecomment-2449439120

Closes #130973.

tbg commented:
> I don't think this would be correct.

Good catch.

> I think it would be unfortunate to reorder these and incur the performance penalty just for test cases like these. Concretely, this would delay requests that are waiting on latches, even when they could have already started running.
> My suggestion would be to remove the Gets in these tests (setupDBAndWriteAAndB, BenchmarkOptimisticEvalForLocks, BenchmarkOptimisticEval) and to run with 1PC + DisableCanAckBeforeReplication instead. The Gets are already a "hack" to wait for completion of the async portions of these writes during setup. Might as well just remove the async work.

Ack, done - see the second commit (first is unchanged I think).

nvb reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 9 of 9 files at r15, 1 of 1 files at r16, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/131071#-:-OB1MirP0pWxpNMgBreY:bnfp4nl)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @tbg)

___
*[`pkg/kv/kvserver/client_replica_test.go` line 4912 at r16](https://reviewable.io/reviews/cockroachdb/cockroach/131071#-OB1MdE70H9oNd-Wyn9g:-OB1MdE70H9oNd-Wyn9h:b-wyich8) ([raw file](https://github.com/cockroachdb/cockroach/blob/5199a8653d9dec3b36539609ead499df34b60059/pkg/kv/kvserver/client_replica_test.go#L4912)):*
> ```Go
> 			}
> 		}()
> 		if err := txn.DisablePipelining(); err != nil {
> ```

Do we need `DisablePipelining` if we're going to `CommitInBatch`?


<!-- Sent from Reviewable.io -->

tbg commented:
> Do we need DisablePipelining if we're going to CommitInBatch?

We shouldn't, removed.

TFTR!

bors r+

miraradeva commented:
blathers backport 24.3
